### PR TITLE
Added domain computation to axes

### PR DESCRIPTION
**Pull request by lorem--ipsum**
Also laid out Options and Dataset classes, I'd like them to handle their own deserialization and validation.

### CODE REVIEW

File: src/LineChart.ts [hunk 1]
```diff
@@ -34,6 +34,7 @@ module n3Charts {
         ['container', Factory.Container, element[0]],
         ['x-axis', Factory.Axis, 'x'],
         ['y-axis', Factory.Axis, 'y'],
+        ['lines-container', Factory.Series.LineSeriesContainer, 'y'],
```
Comment (chaosmail) [inline src/LineChart.ts]: Maybe we could let `Factory.Container` make a `data` g node, and then every `Factory.Series.Line`, `Factory.Series.Area` and so on creates a g node for itself. Like this we could reduce the number of factories.

File: src/LineChart.ts [hunk 2]
```diff
@@ -34,6 +34,7 @@ module n3Charts {
         ['container', Factory.Container, element[0]],
         ['x-axis', Factory.Axis, 'x'],
```
Comment (chaosmail) [inline src/LineChart.ts]: sweet, this could now be `Factory.Axis.SIDE_X` and so on

File: src/LineChart.ts [hunk 3]
```diff
@@ -48,7 +49,7 @@ module n3Charts {
       scope.$watch('options+data', () => {
         // Call the update event with a copy of the options
         // to avoid infinite digest loop
-        eventMgr.trigger('update', scope.datasets, angular.copy(scope.options), attributes);
+        eventMgr.trigger('update', scope.datasets, new Utils.Options(angular.copy(scope.options)), attributes);
```
Comment (chaosmail) [inline src/LineChart.ts]: yep, I was thinking the same. What about the dataset attribute? Should we have also some parser (that removes invisible series from the array or sth? Do we still need to pass the attributes everywhere? Maybe it's ok to just pass them to the options (where the options parser generates valid options out of the attributes).

File: src/factories/Axis.ts [hunk 4]
```diff
@@ -6,11 +6,13 @@ module n3Charts.Factory {
     public svg: D3.Selection;
     public scale: D3.Scale.Scale;
     public axis: D3.Svg.Axis;
+    public static SIDE_X: string = 'x';
```
Comment (chaosmail) [inline src/factories/Axis.ts]: love this. Maybe we can also rewrite the comparisons `=== 'x'` using this static properties

File: src/factories/Axis.ts [hunk 5]
```diff
@@ -30,13 +32,73 @@ module n3Charts.Factory {
...
+        if (series.axis === this.side) {
+          datasetsForSide.push(series.dataset);
+          if (!seriesForDataset[series.dataset]) seriesForDataset[series.dataset] = [];
+          seriesForDataset[series.dataset].push(series);
+        }
+      });
+
+      return this.getExtentForDatasets(
+        datasets,
+        (key) => datasetsForSide.indexOf(key) > -1,
```
Comment (chaosmail) [inline src/factories/Axis.ts]: can we make this a little easier? i mean, i can follow - but I need to jump to the `getExtentForDatasets` method to understand what this is doing. We can still merge this and I can debug a little to propose a simplification.

File: src/factories/series/LineSeriesContainer.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+module n3Charts.Factory.Series {
+  'use strict';
+
+  export class LineSeriesContainer extends Utils.BaseFactory {
```
Comment (chaosmail) [inline src/factories/series/LineSeriesContainer.ts]: This could be already our `Line` factory

File: src/factories/series/LineSeriesContainer.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    public svg: D3.Selection;
+
+    constructor() {
+      super();
+    }
+
+    create() {
+      this.createContainer(this.factoryMgr.get('container').vis);
```
Comment (chaosmail) [inline src/factories/series/LineSeriesContainer.ts]: We could create a data container on the `Factory.Container` and use it here

File: src/factories/series/LineSeriesContainer.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    createContainer(vis:D3.Selection) {
+      this.svg = vis
+        .append('g')
+          .attr('class', 'line-container');
+    }
+
+    update(datasets, options, attributes: ng.IAttributes) {
+    }
```
Comment (chaosmail) [inline src/factories/series/LineSeriesContainer.ts]: I think you need a sth like a comment here to make typescript linter accept it

File: src/utils/Dataset.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    public data: Object[];
+    public key: string;
+
+    constructor(js:any) {
+      if (js) {
+        this.fromJS(js);
+      }
+    }
+
+    fromJS(js:any) {
```
Comment (chaosmail) [inline src/utils/Dataset.ts]: are you planning to parse also other options, like from attributes?

**Comment by chaosmail** [comment]
very nice work. let's discuss a little in the comments before I will merge it

File: src/LineChart.ts (hunk 1 above)
Comment (lorem--ipsum) [inline src/LineChart.ts]: Yeah, I don't really know how to proceed here... I mean if we have the line series directly at this level, we'll end up adding/removing series regarding of the options at this level too... And I don't really want to have a variable number of things...

My idea was too have an entity create as many line series as need, and be just one container from the outside. We would then have the same for any other element type (dots, areas, and any other series type — even weird ones).

File: src/LineChart.ts (hunk 2 above)
Comment (lorem--ipsum) [inline src/LineChart.ts]: Right, I'll change that.

File: src/LineChart.ts (hunk 3 above)
Comment (lorem--ipsum) [inline src/LineChart.ts]: Yeah, I guess we can ignore it for now, and if needed we'll either pass it again or just give it to the options (I prefer the latter).

File: src/factories/Axis.ts (hunk 4 above)
Comment (lorem--ipsum) [inline src/factories/Axis.ts]: Yep, will do !

File: src/factories/Axis.ts (hunk 5 above)
Comment (lorem--ipsum) [inline src/factories/Axis.ts]: Well actually it needs tests... I can write them, but not right now, probably tonight.

File: src/factories/series/LineSeriesContainer.ts (hunk 7 above)
Comment (lorem--ipsum) [inline src/factories/series/LineSeriesContainer.ts]: So the `Factory.Container` would create several `g` ?

File: src/factories/series/LineSeriesContainer.ts (hunk 8 above)
Comment (lorem--ipsum) [inline src/factories/series/LineSeriesContainer.ts]: I have no lint warning... I'll double check that.

File: src/utils/Dataset.ts (hunk 9 above)
Comment (lorem--ipsum) [inline src/utils/Dataset.ts]: Yup, but this needs more tests.

File: src/LineChart.ts (hunk 1 above)
Comment (chaosmail) [inline src/LineChart.ts]: Ok I have the same thing in mind. I will adapt this later and propose a great and simple solution

File: src/LineChart.ts (hunk 3 above)
Comment (chaosmail) [inline src/LineChart.ts]: yes agree, we need to look at this in detail anyway later

File: src/LineChart.ts (hunk 1 above)
Comment (lorem--ipsum) [inline src/LineChart.ts]: :+1:

File: src/factories/series/LineSeriesContainer.ts (hunk 7 above)
Comment (chaosmail) [inline src/factories/series/LineSeriesContainer.ts]: I thought, that the `Factory.Container` should create and provide at least one more g node for all the chart data.
Then we could make the `Factory.Series.Line` that creates another g node inside of the `g.data` to store all line series and so on

Comment (chaosmail) [inline src/factories/series/LineSeriesContainer.ts]: basically just rename `Factory.Series.LineSeriesContainer` to `Factory.Series.Line` and not use the `Factory.Container.vis` property but some data container

Comment (chaosmail) [inline src/factories/series/LineSeriesContainer.ts]: but we could also add this later, haha I love this `dev` branch

**Comment by chaosmail** [comment]
there are some linting errors, but I will still merge - because `dev-branch-yolo`

**Comment by sebastien-fragnaud** [comment]
lol

**Pull request merged**